libsolidity/ast/AST.cpp

```diff
@@ -480,6 +480,66 @@ bool VariableDeclaration::canHaveAutoType() const
 	return isLocalVariable() && !isCallableParameter();
 }
 
+bool VariableDeclaration::isEventParameter() const
+{
+	return dynamic_cast<EventDefinition const*>(scope()) != nullptr;
+}
+
+bool VariableDeclaration::isReferenceType() const
+{
+	return dynamic_cast<ReferenceType const*>(typeName()->annotation().type.get());
+}
+
+set<VariableDeclaration::Location> VariableDeclaration::getAllowedStorageLocations() const
+{
+	using Location = VariableDeclaration::Location;
+
+	if (isReferenceType())
+	{
+		if (isEventParameter())
+			return set<Location>{ Location::Default };
+		else if (isConstant())
+			return set<Location>{ Location::Default, Location::Memory };
+		else if (isExternalCallableParameter())
+		{
+			auto const& contract = dynamic_cast<ContractDefinition const&>(
+					*dynamic_cast<Declaration const&>(*scope()).scope()
+			);
+			if (contract.isLibrary())  // External Library function parameter.
+				return set<Location>{ Location::CallData, Location::Default, Location::Storage };
+			else  // External func function parameter.
+				return set<Location>{ Location::CallData, Location::Default };
```

Comment:
Also no default here.

Reply:
Could you take a look at my 2nd to the last comment above on this? I believe making calldata required would be better suited for a separate PR mainly because of the large number of tests that will have to be updated both here and in OpenZeppelin and in solc-js. I can go ahead and add it all here though if you'd rather me.